src/test/ui/iterators/issue-58952-filter-type-length.rs

```diff
@@ -27,5 +28,5 @@ fn main() {
         .filter(|a| b.clone().any(|b| *b == *a))
         .filter(|a| b.clone().any(|b| *b == *a))
         .filter(|a| b.clone().any(|b| *b == *a))
-        .collect::<Vec<_>>();
+        .collect::<VecDeque<_>>();
```

Comment:
CC @lcnr 

Based on the original issue this was an artificial test-case. What exactly is it supposed to test? Specifically, is the case of sourcing things from a vec and collecting into a vec relevant here?

This PR changes from external to internal iteration for collect, which does increase the type length, but not significantly more so than you would already have experienced if you had used another method using internal iteration such as `.count()`

Reply:
I added this test to guard against unintentional blowup here,
as we previously could cause an increased type length without noticing it.

Feel free to adapt this test if the increase isn't too large and expected.